### PR TITLE
Pipeline compatibility

### PR DESCRIPTION
**Pull request by lugowskii**

### CODE REVIEW

File: pipeline-steps-explained.MD [hunk 1]
```diff
@@ -0,0 +1,93 @@
+Dynatrace AppMon Pipeline steps
+-------------------------------
+Requirements: Jenkins 2.0+ with <a href="https://wiki.jenkins.io/display/JENKINS/Pipeline+Plugin">Pipeline Plugin</a> is supported. Older versions could work, but are no officially supported.
```
Comment (greg-witczak) [inline pipeline-steps-explained.MD]: "not officially supported"

File: pipeline-steps-explained.MD [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+Warning: If you are using Jenkins in version lower than 2.7.3(2016/08/31), please ensure you have Pipeline: Nodes and Processes plugin in version at most 2.13.
+
+Warning: To have 'Test result trend' and 'Latest Test Result' links on Maven project main page, please use <a href="https://wiki.jenkins.io/display/JENKINS/Maven+Project+Plugin">Maven Integration Plugin</a> in version 2.13+.
+
+Global configuration is the same as for freestyle projects. It's described [here](README.MD) in Global Settings section.
+> When not specified, assume step parameter is literal type.
+
+---
+#### **appMonBuildEnvironment**
+Corresponds to 'Use Dynatrace AppMon to monitor tests' from 'Build environment' section. It is a wrapper - all other appMon steps should be surrounded by `appMonBuildEnvironment`.
```
Comment (greg-witczak) [inline pipeline-steps-explained.MD]: "appMon" -> "AppMon"

File: pipeline-steps-explained.MD [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+---
+#### **appMonRegisterTestRun**
+Corresponds to 'Dynatrace AppMon - Register Test Run' step from freestyle projects.
+
+* Required parameter:
+    * `category` (unit, uidriven, performance, webapi, external)
+* Optional parameter:
+    * `platform`
+
+Returns registered testRunId, which must be injected into test execution.
```
Comment (greg-witczak) [inline pipeline-steps-explained.MD]: Maybe add "code" formatting for `testRunId` (using ` char)

File: pipeline-steps-explained.MD [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+##### Example of usage (with Maven):
+With Maven and Surefire or Failsafe plugin you don't need any additional configuration in your project. Everything is done from jenkins pipeline script.
```
Comment (greg-witczak) [inline pipeline-steps-explained.MD]: "you don't need any additional configuration" -> "you don't need additional configuration"

File: src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java [hunk 5]
```diff
@@ -170,16 +193,17 @@ private void setupBuildVariables(AbstractBuild build, String serverUrl) {
...
 		}
-		if (StringUtils.isNotEmpty(globalConfig.password)) {
+		if (globalConfig != null && StringUtils.isNotEmpty(globalConfig.password)) {
 			parameters.add(new PasswordParameterValue(BuildVarKeys.BUILD_VAR_KEY_GLOBAL_PASSWORD, globalConfig.password));
 		}
 		Utils.updateBuildVariables(build, parameters);
 	}
 
 	@Extension
+	@Symbol("appMonBuildEnvironment")/**is a function name in pipeline script*/
```
Comment (greg-witczak) [inline src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java]: If it's non-javadoc comment, use `/* */`
If it's javadoc comment, write it in new line, above annotations.

File: src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        private transient Run<?, ?> build;
+
+        @Override
+        protected String run() throws Exception {
+            TATestRunRegistrationBuildStep buildStep = new TATestRunRegistrationBuildStep(step.category, step.platform);
+            buildStep.perform(build, workspace, launcher, listener);
+            if (buildStep.getTestRunId() != null) {
+                return buildStep.getTestRunId();
+            }
+            else{
```
Comment (greg-witczak) [inline src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java]: `else` is unnecessary here

File: src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            super(Execution.class);
+        }
+
+        public static String getDefaultCategory() {
+            return DEFAULT_CATEGORY;
+        }
+
+        @Override
+        public String getFunctionName() {
+            return "appMonRegisterTestRun";
```
Comment (greg-witczak) [inline src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java]: Is is possible to extract it to constant, somewhere close to actual method name?

File: src/main/resources/com/dynatrace/jenkins/dashboard/steps/TAReportingRecorderPipelineStep/help.html [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<div>
+  This plugin fetches analysis data from the <a href="http://www.dynatrace.com">Dynatrace Server</a> (Test Center Edition) REST
```
Comment (greg-witczak) [inline src/main/resources/com/dynatrace/jenkins/dashboard/steps/TAReportingRecorderPipelineStep/help.html]: 1) AppMon server
2) Change url to AppMon doc: https://www.dynatrace.com/support/doc/appmon/

File: pipeline-steps-explained.MD [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+##### Example of usage (with Maven):
+With Maven and Surefire or Failsafe plugin you don't need additional configuration in your project. Everything is done from jenkins pipeline script.
+
+```groovy
+def testRunId = appMonRegisterTestRun category: 'unit'
+bat(<your Maven path> -DargLine="-agentpath:<path to dtagent.dll/dtagent.so>=name=<agent name>,server=<host[:port]>,loglevel=warning,optionTestRunIdJava=${testRunId}" <your Maven goals>/)
+```
+
+
+##### Example of usage (with Ant):
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: example of usage -> Usage example

File: pipeline-steps-explained.MD [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+Corresponds to 'Dynatrace AppMon - Register Test Run' step from freestyle projects.
+
+* Required parameter:
+    * `category` (unit, uidriven, performance, webapi, external)
+* Optional parameter:
+    * `platform`
+
+Returns registered `testRunId`, which must be injected into test execution.
+##### Example of usage (with Maven):
+With Maven and Surefire or Failsafe plugin you don't need additional configuration in your project. Everything is done from jenkins pipeline script.
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: Using Surefire or Failsafe plugin there's no need to adapt your project's configuration, except for few Jenkins pipeline script changes described below.

File: pipeline-steps-explained.MD [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+```groovy
+def testRunId = appMonRegisterTestRun category: 'unit'
+bat(<your Maven path> -DargLine="-agentpath:<path to dtagent.dll/dtagent.so>=name=<agent name>,server=<host[:port]>,loglevel=warning,optionTestRunIdJava=${testRunId}" <your Maven goals>/)
+```
+
+
+##### Example of usage (with Ant):
+
+With Ant it is needed to add few lines to build script, as described [here](example-ant-with-jenkins.md).
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: Follow the instructions described [here](example-ant-with-jenkins.md) and adapt your pipeline script as follows:

File: pipeline-steps-explained.MD [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+bat(<your Ant path> -DdtTestrunID=${testRunId} <Ant targets>/)
+```
+---
+#### **appMonPublishTestResults**
+Corresponds to 'Dynatrace AppMon - Publish Test Results' post-build step from freestyle projects.
+
+* No required parameters.
+* Optional parameters:
+    * `modifyStatusIfDegraded`
+    * `statusNameIfDegraded`
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: should be possible to pick one of supported values only
is it possible to combine these two options into one?
when user provides statusNameIfDegraded other than null/empty then it's obvious that he wants to change that status

File: pipeline-steps-explained.MD [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    * `modifyStatusIfVolatile`
+    * `statusNameIfVolatile`
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: should be possible to pick one of supported values only
is it possible to combine these two options into one?
when user provides statusNameIfVolatile other than null/empty then it's obvious that he wants to change that status

File: pipeline-steps-explained.MD [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+```groovy
+def testRunId = appMonRegisterTestRun category: 'unit'
+bat(<your Ant path> -DdtTestrunID=${testRunId} <Ant targets>/)
+```
+---
+#### **appMonPublishTestResults**
+Corresponds to 'Dynatrace AppMon - Publish Test Results' post-build step from freestyle projects.
+
+* No required parameters.
+* Optional parameters:
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: describe params

File: pipeline-steps-explained.MD [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    * `modifyStatusIfDegraded`
+    * `statusNameIfDegraded`
+    * `modifyStatusIfVolatile`
+    * `statusNameIfVolatile`
+    * `printXmlReportForDebug` [boolean]
+
+##### Examples of usage:
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: examples of usage -> Usage examples

File: pipeline-steps-explained.MD [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+```groovy
+appMonPublishTestResults()
+```
+With all parameters:
+```groovy
+appMonPublishTestResults modifyStatusIfDegraded: true, modifyStatusIfVolatile: true, printXmlReportForDebug: true, statusNameIfDegraded: 'UNSTABLE', statusNameIfVolatile: 'UNSTABLE'
+```
+---
+#### Pipeline tip:
+If you are not familiar with pipeline syntax, you could go to your pipeline project menu and choose 'Pipeline syntax'. It's the best place to begin work with pipelines in Jenkins.
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: could -> can
best place to begin work -> best place to start

File: src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java [hunk 17]
```diff
@@ -61,7 +66,7 @@
 /**
  * Created by krzysztof.necel on 2016-02-09.
  */
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java]: add yourself as an @author

File: src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java [hunk 18]
```diff
@@ -170,16 +193,17 @@ private void setupBuildVariables(AbstractBuild build, String serverUrl) {
 		if (StringUtils.isNotEmpty(serverUrl)) {
 			parameters.add(new StringParameterValue(BuildVarKeys.BUILD_VAR_KEY_GLOBAL_SERVER_URL, serverUrl));
 		}
-		if (StringUtils.isNotEmpty(globalConfig.username)) {
+		if (globalConfig != null && StringUtils.isNotEmpty(globalConfig.username)) {
 			parameters.add(new StringParameterValue(BuildVarKeys.BUILD_VAR_KEY_GLOBAL_USERNAME, globalConfig.username));
 		}
-		if (StringUtils.isNotEmpty(globalConfig.password)) {
+		if (globalConfig != null && StringUtils.isNotEmpty(globalConfig.password)) {
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java]: change to:
```
if(globalConfig != null) {
  if(username not empty) ...
  if(password not empty) ...
}
```

File: src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java [hunk 19]
```diff
@@ -170,16 +193,17 @@ private void setupBuildVariables(AbstractBuild build, String serverUrl) {
...
 			parameters.add(new PasswordParameterValue(BuildVarKeys.BUILD_VAR_KEY_GLOBAL_PASSWORD, globalConfig.password));
 		}
 		Utils.updateBuildVariables(build, parameters);
 	}
 
 	@Extension
+	@Symbol("appMonBuildEnvironment")/*is a function name in pipeline script*/
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/TABuildWrapper.java]: any participant needing to change something here should be familiar with the @Symbol annotation and the parameter meaning, no need to document that

File: src/main/java/com/dynatrace/jenkins/dashboard/TAReportingBuildAction_2_0_0.java [hunk 20]
```diff
@@ -32,27 +32,31 @@
...
 import java.net.URISyntaxException;
 import java.text.MessageFormat;
+import java.util.Collection;
+import java.util.Collections;
 
 /**
  * Created by krzysztof.necel on 2016-02-05.
  */
-public class TAReportingBuildAction_2_0_0 implements Action, StaplerProxy {
+public class TAReportingBuildAction_2_0_0 implements Action, StaplerProxy, SimpleBuildStep.LastBuildAction, RunAction2{
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/TAReportingBuildAction_2_0_0.java]: check your formatter, opening brackets should be preceded with a space
fix all the occurences

File: src/main/java/com/dynatrace/jenkins/dashboard/TATestRunRegistrationBuildStep.java [hunk 21]
```diff
@@ -107,19 +113,18 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
...
-			return true;
 		} catch (Exception e) {
 			e.printStackTrace();
 			logger.println("ERROR: Dynatrace AppMon Plugin - build step execution failed (see the stacktrace to get more information):\n" + e.toString());
-			return false;
 		}
 	}
 
-	private void updateTestRunIdsAction(AbstractBuild build, String newTestRunId) {
+	private void updateTestRunIdsAction(Run build, String newTestRunId) {
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/TATestRunRegistrationBuildStep.java]: rename build -> run (same for all similar changes)

File: src/main/java/com/dynatrace/jenkins/dashboard/model_2_0_0/TAReport.java [hunk 22]
```diff
@@ -41,16 +41,16 @@
  */
 public class TAReport {
 
-	private final AbstractBuild<?, ?> build;
+	private final Run<?, ?> build;
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/model_2_0_0/TAReport.java]: rename build -> run

File: src/main/java/com/dynatrace/jenkins/dashboard/model_2_0_0/TAReport.java [hunk 23]
```diff
@@ -41,16 +41,16 @@
...
 	private transient SoftReference<TAReportDetails> reportDetailsRef;
 
-	public TAReport(Map<TestStatus, Integer> summary, AbstractBuild<?, ?> build) {
+	public TAReport(Map<TestStatus, Integer> summary, Run<?, ?> build) {
 		this.summary = summary;
 		this.build = build;
 	}
 
-	public AbstractBuild<?, ?> getBuild() {
+	public Run<?, ?> getBuild() {
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/model_2_0_0/TAReport.java]: rename to getRun()

File: src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+import hudson.util.ListBoxModel;
+import org.jenkinsci.Symbol;
+import org.jenkinsci.plugins.workflow.steps.*;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+
+/**Extracted to new step, because we need to return a value (testRunId) from step*/
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java]: explanations like that aren't javadocs, but simple (single- or multiline) comments

File: src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java (hunk 7 above)
Comment (dglugla) [reply src/main/java/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep.java]: +1

File: src/main/java/com/dynatrace/jenkins/dashboard/utils/Utils.java [hunk 25]
```diff
@@ -61,15 +61,18 @@ private Utils() {
...
-		BasicServerConfiguration config = new BasicServerConfiguration(globalConfig.username,
-				globalConfig.password,
-				globalConfig.protocol.startsWith("https"),
-				globalConfig.host,
-				globalConfig.port,
-				globalConfig.validateCerts,
-				//connection timeout, 0 stands for infinite
-				0);
-		return new DynatraceClient(config);
+		if (globalConfig != null) {
```
Comment (dglugla) [inline src/main/java/com/dynatrace/jenkins/dashboard/utils/Utils.java]: check your formatter - the code looks ugly now, but maybe GH makes some formatting while rendering the page

File: src/main/resources/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep/help.html [hunk 26]
```diff
@@ -0,0 +1,5 @@
+<div>
+  <p>
+    This step registers new test run with given category and returns test run id
```
Comment (dglugla) [inline src/main/resources/com/dynatrace/jenkins/dashboard/steps/TATestRunRegistrationPipelineStep/help.html]: capitalize "Test Run"

File: src/test/java/com/dynatrace/jenkins/dashboard/utils/UtilsCompatTest.java [hunk 27]
```diff
@@ -37,7 +38,7 @@ public void getCompatReportTest() {
 		TestAutomationBuildAction buildAction = mock(TestAutomationBuildAction.class);
 		when(buildAction.getTestAutomationReport()).thenReturn(oldReport);
 
-		AbstractBuild build = mock(AbstractBuild.class);
+		Run build = mock(Run.class);
```
Comment (dglugla) [inline src/test/java/com/dynatrace/jenkins/dashboard/utils/UtilsCompatTest.java]: build -> run

File: src/test/java/com/dynatrace/jenkins/dashboard/utils/UtilsCompatTest.java [hunk 28]
```diff
@@ -47,7 +48,7 @@ public void getCompatReportTest() {
 		assertThat("Result summary must not be null", result, is(notNullValue()));
 		TestUtils.assertSummary(resultSummary, 1, 2, 3, 4, 0, 0);
 
-		AbstractBuild<?, ?> resultBuild = result.getBuild();
+		Run<?, ?> resultBuild = result.getBuild();
```
Comment (dglugla) [inline src/test/java/com/dynatrace/jenkins/dashboard/utils/UtilsCompatTest.java]: same here

File: pipeline-steps-explained.MD [hunk 29]
```diff
@@ -0,0 +1,93 @@
...
+#### **appMonBuildEnvironment**
+Corresponds to 'Use Dynatrace AppMon to monitor tests' from 'Build environment' section. It is a wrapper - all other `appMon*` steps should be surrounded by `appMonBuildEnvironment`.
+
+* Required parameter:
+    * `systemProfile`
+* Optional parameters:
+    * `versionMajor`
+    * `versionMinor`
+    * `versionRevision`
+    * `versionMilestone`
```
[the remark this answers is not in the document]
Comment (lugowskii) [reply pipeline-steps-explained.MD]: As far as I know there is no option for user to pass versionBuild in Build Environment section.

[the remark this answers is not in the document]
Comment (lugowskii) [reply pipeline-steps-explained.MD]: versionBuild  is set to build number.
TaTestEunRegistrationBuildStep:96 ` versionBuild = Integer.toString(run.getNumber());`

File: src/main/java/com/dynatrace/jenkins/dashboard/utils/Utils.java (hunk 25 above)
Comment (lugowskii) [reply src/main/java/com/dynatrace/jenkins/dashboard/utils/Utils.java]: I think it's Github formatting.

File: src/main/java/com/dynatrace/jenkins/dashboard/model_2_0_0/TAReport.java (hunk 22 above)
Comment (lugowskii) [reply src/main/java/com/dynatrace/jenkins/dashboard/model_2_0_0/TAReport.java]: Renaming these variables requires a lot of changes in .jelly files .

File: pipeline-steps-explained.MD [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+Global configuration remains unchanged and is described [here](README.MD) in Global Settings section.
+
+---
+#### **appMonBuildEnvironment**
+Corresponds to 'Use Dynatrace AppMon to monitor tests' from 'Build environment' section. Wrapper that provides constant Test Run metadata for all 'appMonRegisterTestRun' steps defined.
+
+* Required parameter:
+    * `systemProfile`
+* Optional parameters:
+    * `version*` - version descriptors aggregated to "major.minor.revision.milestone" pattern
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: you have to point all the params here: versionMajor, versionMinor, and so on; user needs to know what are their names
actually we aggregate to 'major.minor.revision **.build** milestone' in AppMon, even though the versionBuild is set automatically

File: pipeline-steps-explained.MD [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+appMonBuildEnvironment(marker: '1', recordSession: true, systemProfile: 'test_jenkins', versionMajor: '5', versionMilestone: '2', versionMinor: '4', versionRevision: '3') {
+    //your pipeline logic
+}
+```
+---
+#### **appMonRegisterTestRun**
+Has the same purpose as 'Dynatrace AppMon - Register Test Run' step - registers a Test Run used to differentiate subsets of tests. Uses Test Run's metadata defined in 'appMonBuildEnvironment' wrapper step.
+
+* Required parameter:
+    * `category` - category of TestRun(unit, uidriven, performance, webapi, external)
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: space before '('
remove 'external' category, it's designed for a different purposes, and IMHO shouldn't be used this way

File: pipeline-steps-explained.MD (hunk 13 above)
Comment (dglugla) [reply pipeline-steps-explained.MD]: what about this one and the previous one?

File: pipeline-steps-explained.MD [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+#### **appMonBuildEnvironment**
+Corresponds to 'Use Dynatrace AppMon to monitor tests' from 'Build environment' section. Wrapper that provides constant Test Run metadata for all 'appMonRegisterTestRun' steps defined.
+
+* Required parameter:
+    * `systemProfile`
+* Optional parameters:
+    * `versionMajor` - 	"A" section of "A.B.C.D Milestone E" version ('dtVersionMajor' var)
+    * `versionMinor` - "B" section of "A.B.C.D Milestone E" version ('dtVersionMinor' var)
+    * `versionRevision` - "C" section of "A.B.C.D Milestone E" version ('dtVersionRevision' var)
+    * `versionMilestone` - "E" section of "A.B.C.D Milestone E" version ('dtVersionMilestone' var)
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: I was thinking more of something like:
* 'versionMajor', 'versionMinor', 'versionRevision', 'versionMilestone' - version descriptors aggregated to 'major.minor.revision.build milestone' pattern, where the build number is supplied by Jenkins
in that case - wihout the note below (about aggregated pattern)

File: pipeline-steps-explained.MD [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+    * `marker` - string literal to provide a user-defined marker
+    * `recordSession` - boolean flag to enable session recording
+
+    Version descriptors are aggregated to 'major.minor.revision.build milestone' pattern
+
+##### Examples of usage:
```
Comment (dglugla) [inline pipeline-steps-explained.MD]: Usage example

**Comment by dglugla** [comment]
I'm also missing some link to the pipeline documentation on the main page

File: README.md [hunk 34]
```diff
@@ -4,6 +4,8 @@
 
 This plugin for Jenkins pulls Test Automation data from Dynatrace AppMon and displays it through charts and tables on the project and build level.
 
+Since version 2.1.0 plugin supports Jenkins Pipelines. Required configuration and usage examples are described [here](pipeline-steps-explained.md). Jenkins 2.0 and later is officially supported.
```
Comment (dglugla) [inline README.md]: Release 2.1.0 adds support for Jenkins Pipelines. Configuration and usage examples are described [here](pipeline-steps-explained.md).

**Comment by lugowskii** [comment]
- 3 pipeline steps created
- Descriptions for steps modified to adhere to pipeline specifics
- code refactored according to Jenkins plugin developer pipeline
Ticket JLT-164950

**Pull request merged**